core/lib/dal/src/blocks_dal.rs

```diff
@@ -281,7 +283,18 @@ impl BlocksDal<'_, '_> {
         .fetch_all(self.storage)
         .await?;
 
-        Ok(l1_batches.into_iter().map(Into::into).collect())
+        let mut l1_batch_headers = Vec::with_capacity(storage_l1_batch_headers.len());
+
+        for batch in storage_l1_batch_headers {
+            let l2_to_l1_logs = self
+                .storage
+                .blocks_web3_dal()
+                .get_l2_to_l1_logs_by_number::<UserL2ToL1Log>(L1BatchNumber(batch.number as u32))
+                .await?;
+            l1_batch_headers.push(batch.into_l1_batch_header_with_logs(l2_to_l1_logs));
```

Comment:
Are there supposed to be a few L1 batches requested here? Otherwise, it may make sense to generalize `get_l2_to_l1_logs_by_number()` to support a set (range?) of batches. OTOH, `get_l1_batches_for_eth_tx_id()` seems to only be used in [metrics for `eth_sender`](https://github.com/matter-labs/zksync-era/blob/298a97e800b4c156628050789de7a490a7565d60/core/node/eth_sender/src/metrics.rs#L135) that doesn't really needs logs or most of other L1 batch header fields (it only needs timestamps and transaction counts), so it may make sense to refactor it and remove this method altogether.

Reply:
>, so it may make sense to refactor 

By looking quickly into [metrics for eth_sender](https://github.com/matter-labs/zksync-era/blob/298a97e800b4c156628050789de7a490a7565d60/core/node/eth_sender/src/metrics.rs#L135)..the only quick way for refactoring I see is to *modify* `get_l1_batches_for_eth_tx_id()`, by asking db only for timestamps, l1_tx_count and l2_tx_count(Idk it will improve smth)
To get L1 batches' related info(timestamps, l1_tx_count and l2_tx_count), having only EthTx will be at least not easy/quick.

Reply:
Sorry, that's what I meant, but I wasn't specific enough: create a method like this

```rust
pub struct L1BatchStatistics {
    pub number: L1BatchNumber,
    pub timestamp: u64,
    pub transaction_count: u32,
}

pub async fn get_l1_batch_statistics_for_eth_tx_id(
    &mut self,
    eth_tx_id: u32,
) -> DalResult<Vec<L1BatchStatistics>> {
    // ...
}
```

Reply:
I see! Now makes much more sense for me!